src/gpuarray_buffer_cuda.c

```diff
@@ -49,18 +49,72 @@ static int cuda_records(gpudata *, int, CUstream);
 static int detect_arch(const char *prefix, char *ret, CUresult *err);
 static gpudata *new_gpudata(cuda_context *ctx, CUdeviceptr ptr, size_t size);
 
-static int strb_eq(void *_k1, void *_k2) {
-  strb *k1 = (strb *)_k1;
-  strb *k2 = (strb *)_k2;
+typedef struct _kernel_key {
+  char bin_id[64];
+  strb *src;
+} kernel_key;
+
+static void key_free(cache_key_t _k) {
+  kernel_key *k = (kernel_key *)_k;
+  strb_free(k->src);
+  free(k);
+}
+
+static int strb_eq(strb *k1, strb *k2) {
   return (k1->l == k2->l &&
           memcmp(k1->s, k2->s, k1->l) == 0);
 }
 
-static uint32_t strb_hash(void *_k) {
-  strb *k = (strb *)_k;
+static uint32_t strb_hash(strb *k) {
   return XXH32(k->s, k->l, 42);
 }
 
+static int key_eq(kernel_key *k1, kernel_key *k2) {
+  return (memcmp(k1->bin_id, k2->bin_id, 64) == 0 &&
+          strb_eq(k1->src, k2->src));
+}
+
+static int key_hash(kernel_key *k) {
+  XXH32_state_t state;
+  XXH32_reset(&state, 42);
+  XXH32_update(&state, k->bin_id, 64);
+  XXH32_update(&state, k->src->s, k->src->l);
+  return XXH32_digest(&state);
+}
+
+static int key_write(strb *res, kernel_key *k) {
+  strb_appendn(res, k->bin_id, 64);
+  strb_appendb(res, k->src);
+  return strb_error(res);
+}
+
+static kernel_key *key_read(const strb *b) {
+  kernel_key *k;
+  if (b->l < 64) return NULL;
+  k = malloc(sizeof(*k));
+  if (k == NULL) return NULL;
+  k->src = strb_alloc(b->l - 64);
+  if (k->src == NULL) {
+    free(k);
+    return NULL;
+  }
+  memcpy(k->bin_id, b->s, 64);
+  strb_appendn(k->src, b->s+64, b->l-64);
+  return k;
+}
+
+static int kernel_write(strb *res, strb *bin) {
+  strb_appendb(res, bin);
+  return strb_error(res);
+}
+
+static strb *kernel_read(const strb *b) {
```

Comment:
In this case it does more or less a copy.  The point is that it has to fit into an API that asks it to read data from the input strb and deserialize the data.  It just happens that the data in this case is another strb.